iceoryx_posh/source/popo/ports/server_port_roudi.cpp

```diff
@@ -38,19 +39,117 @@ ServerPortRouDi::MemberType_t* ServerPortRouDi::getMembers() noexcept
 }
 
 cxx::optional<capro::CaproMessage>
-ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& /*caProMessage*/) noexcept
+ServerPortRouDi::dispatchCaProMessageAndGetPossibleResponse(const capro::CaproMessage& caProMessage) noexcept
 {
-    capro::CaproMessage responseMessage(
-        capro::CaproMessageType::NACK, this->getCaProServiceDescription(), capro::CaproMessageSubType::NOSUBTYPE);
+    const auto isOffered = getMembers()->m_offered.load(std::memory_order_relaxed);
+
+    return isOffered ? handleCapProMessageForStateOffered(caProMessage)
+                     : handleCapProMessageForStateNotOffered(caProMessage);
+}
 
-    return cxx::make_optional<capro::CaproMessage>(responseMessage);
+void ServerPortRouDi::handleCaProProtocolViolation(const capro::CaproMessageType messageType) const noexcept
+{
+    // this shouldn't be reached
+    LogFatal() << "CaPro Protocol Violation! Got '" << messageType << "' with offer state '"
+               << (getMembers()->m_offered ? "OFFERED" : "NOT OFFERED") << "'!";
+    errorHandler(Error::kPOPO__CAPRO_PROTOCOL_ERROR, nullptr, ErrorLevel::SEVERE);
+}
+
+cxx::optional<capro::CaproMessage>
+ServerPortRouDi::handleCapProMessageForStateOffered(const capro::CaproMessage& caProMessage) noexcept
+{
+    capro::CaproMessage responseMessage{capro::CaproMessageType::NACK, this->getCaProServiceDescription()};
+
+    switch (caProMessage.m_type)
+    {
+    case capro::CaproMessageType::STOP_OFFER:
+        getMembers()->m_offered.store(false, std::memory_order_relaxed);
+        m_chunkSender.removeAllQueues();
+        return caProMessage;
+    case capro::CaproMessageType::OFFER:
+        return responseMessage;
+    case capro::CaproMessageType::CONNECT:
+        if (caProMessage.m_chunkQueueData == nullptr)
+        {
+            LogWarn() << "No client response queue passed to server";
+            errorHandler(Error::kPOPO__SERVER_PORT_NO_CLIENT_RESPONSE_QUEUE_TO_CONNECT, nullptr, ErrorLevel::MODERATE);
+        }
+        else
+        {
+            m_chunkSender
+                .tryAddQueue(static_cast<ClientChunkQueueData_t*>(caProMessage.m_chunkQueueData),
+                             caProMessage.m_historyCapacity)
+                .and_then([this, &responseMessage]() {
+                    responseMessage.m_type = capro::CaproMessageType::ACK;
+                    responseMessage.m_chunkQueueData = static_cast<void*>(&getMembers()->m_chunkReceiverData);
+                    responseMessage.m_historyCapacity = 0;
+                });
+        }
+        return responseMessage;
+    case capro::CaproMessageType::DISCONNECT:
+        m_chunkSender.tryRemoveQueue(static_cast<ClientChunkQueueData_t*>(caProMessage.m_chunkQueueData))
+            .and_then([&responseMessage]() { responseMessage.m_type = capro::CaproMessageType::ACK; });
+        return responseMessage;
+    default:
+        break;
+    }
+
+    handleCaProProtocolViolation(caProMessage.m_type);
```

Comment:
Could you please move this into the `default` case like in the method below. It is a bit more consistent and clear that when no value fits its automatically a protocol violation.

Reply:
Then I would still need to have it here since there are cases where a branch does not return and leaves the switch statement. Alternatively I would need to ensure that all branches return which is also error prone when new branches are added. I'd rather move the `handleCaProProtocolViolation` out the the default case from the method below. What do you think?